### PR TITLE
#167313420 A user should be able to see their reading statistics

### PR DESCRIPTION
**Pull request by Anguandia**
#### What does this PR do?

  Add user reading statistics feature

#### Description of Task to be completed?

  - Check every get request for article resources for authorisation headers
  - Log every resource resource got by authenticated user just before requested resource is successfully returned
  - Save to stats table with item type as article or comment
  - Return user stats on request to owning user

#### How should this be manually tested?

  - Signin and place a GET/users/stats request
  - Then place some GET/articles/:slug requests and get back to the stats request
    - Note the change in response of the GET/users/stats request

#### What are the relevant pivotal tracker stories?

  [WIP #167313420](https://www.pivotaltracker.com/story/show/167313420)

#### Screenshots (if appropriate)

Successful stats fetch
Screen Shot 2019-08-29 at 12 12 56

Un authenticated request
Screen Shot 2019-08-29 at 12 15 57

Response when no statistics for current user
Screen Shot 2019-08-29 at 12 18 43

### CODE REVIEW

File: src/routes/api/article/article.routes.js [hunk 1]
```diff
@@ -22,7 +23,7 @@ const {
 router.post('/:articleId/favorite', [auth, confirmEmailAuth, validateId], FavoritesController.createOrRemoveFavorite);
 router.post('/', [auth, confirmEmailAuth], imageUpload.array('images', 10), validate(schema.articleSchema), articleController.createArticles);
 router.get('/', checkQuery, articleController.getAllArticles);
-router.get('/:slug', articleController.getOneArticle);
+router.get('/:slug', StatsWare.saveStat, articleController.getOneArticle);
```
Comment (salviosage) [inline src/routes/api/article/article.routes.js:28]: why did you change this? any user(logged in or not) can get a single article .

File: src/middlewares/stats.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import jwt from 'jsonwebtoken';
+
+import StatsService from '../services/db.service';
+import { notFound, error } from '../helpers/responses';
+
+const { getStat } = StatsService;
+
+class statsWare {
+  static async checkStats(req, res, next) {
```
Comment (salviosage) [inline src/middlewares/stats.js:11]: can you please use auth middleware

**Review by salviosage — CHANGES_REQUESTED** [review]

File: src/middlewares/stats.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        req.auth = decode;
+        next();
+      });
+    } catch (err) {
+      next();
+    }
+  }
+}
+
+export default statsWare;
```
Comment (minega25) [inline src/middlewares/stats.js:37]: What does the statsWare middleware do @Anguandia

File: src/routes/api/article/article.routes.js (hunk 1 above)
Comment (minega25) [inline src/routes/api/article/article.routes.js:28]: Can an Unauthenticated user get his reading statistics?

File: src/migrations/20190828152317-create-stats.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        type: Sequelize.INTEGER
+      },
+      slug: {
+        type: Sequelize.STRING
+      },
+      item: {
+        type: Sequelize.STRING
+      },
+      readerId: {
+        type: Sequelize.INTEGER
```
Comment (minega25) [inline src/migrations/20190828152317-create-stats.js:18]: What happens if the user is deleted?

File: src/models/stats.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+  const Stats = sequelize.define('Stats', {
+    slug: DataTypes.STRING,
+    item: DataTypes.STRING,
+    readerId: DataTypes.INTEGER
```
Comment (minega25) [inline src/models/stats.js:6]: In my opinion I think we can have a relationship between readerId and user table. What do you think @Anguandia

File: src/routes/api/article/article.routes.js (hunk 1 above)
Comment (Anguandia) [reply src/routes/api/article/article.routes.js:28]: That middle ware is transparent, doesn't return any responses but only scans for the auth header to ascertain loginstatus of requesting user. Couldn't put auth in that route because it is not supposed to be protected

File: src/middlewares/stats.js (hunk 2 above)
Comment (Anguandia) [reply src/middlewares/stats.js:11]: For the same reason above, the auth middleware as is will interrupt requests and return auth responses. All we need here is a transparent check

File: src/middlewares/stats.js (hunk 3 above)
Comment (Anguandia) [reply src/middlewares/stats.js:37]: Thanks for the question Patrick. 
  - On the get article route, the saveStat method of statsWare does the scanning for the auth header in the request and if found, decodes the token and appends an auth property to the request so that the getOne method in the article controller can extract the current userId from this auth property for constituting this requests statistic.
  - The same function in the users/getstats route identifies the user whose statistics should be fetched; pretty much like the auth function but will more streamlined to tis particular function, only reminding the requester to login if not done
  - The checkStat method of the statsWare middleware will check the stats table for user stats after the saveStat confirms authenticated user. If not found, will return the 404 response before the request hits the stats controller. In this way, the stats controller is only left with processing and returning responses for confirmed viable stats get requests

File: src/routes/api/article/article.routes.js (hunk 1 above)
Comment (Anguandia) [reply src/routes/api/article/article.routes.js:28]: No Patric, it is the user stats, and the only way of identifying the user is through his auth token/userid or email extracted from here

**Review by Mnickii — APPROVED** [review]
LGTM

**Review by minega25 — APPROVED** [review]

File: src/models/stats.js (hunk 5 above)
Comment (Anguandia) [reply src/models/stats.js:6]: Yes, readerId actually translates to the id in the user table

**Review by salviosage — APPROVED** [review]
